### ISSUE TITLE
Using offset storage in Kafka and not Zookeeper not working

### ISSUE THREAD

**Issue by ronhash10** [pinterest/secor#447]
Hey,
We are trying to move from Zookeeper to Kafka offset storage:
```
kafka.offsets.storage=kafka
kafka.dual.commit.enabled=false
```
We see though, that two consumers are created, one in Zookeeper, and one in Kafka, but the one that secor uses to commit and read the offset from is the Zookeeper.
Is there any support for having offset storage in Kafka in Secor?
Thanks,
Ron.

**Comment by HenryCaiHaiying** [comment]
you need to set two params, take a look at: https://github.com/pinterest/secor/commit/c3c9b2dc745b6be858e070ff5d8d04f768d11ba9#diff-5bb7557074535d902a0519f1c5faeb23

**Comment by HenryCaiHaiying** [comment]
I feel secor will always use zookeeper unless the code in ZookeeperConnector changes.  You can probably use dual commit to commit the offsets to both __consumer_offsets topic and ZK

**Comment by sjoerdmulder** [comment]
We are facing the same issue; it seems that the kafka offset storage is not working and not used; It's only creating a zookeeper one. Which could also be seen here: https://github.com/pinterest/secor/blob/5d2fc90d6725f6abf8cb0acdbfc9f23187d3a131/src/main/java/com/pinterest/secor/uploader/Uploader.java#L114

**Comment by HenryCaiHaiying** [comment]
I don't think secor is going to read from kafka offset storage unless these
code be changed.

But on the other hand, even if you moved offset storage to kafka, the ZK
dependency will still be there (kafka brokers will still need ZK to store
lots of other metadata information).

**Comment by juhoautio** [comment]
I also tried migrating from zookeeper to kafka offsets storage, but got data loss. How can I migrate the offset storage without losing any data?

Secor seems to commit offsets to both kafka & zookeeper now (all partitions), but file uploading seems to be off sync. It could be related to this condition:

https://github.com/pinterest/secor/blob/c457ed4df255d5dcc9ce78ab6bd07f7faec3d08b/src/main/java/com/pinterest/secor/uploader/Uploader.java#L113-L116

----

Previous conf:

```properties
kafka.dual.commit.enabled=false
kafka.offsets.storage=zookeeper
```

Changed like this and restarted secor:

```properties
kafka.dual.commit.enabled=true
kafka.offsets.storage=kafka
```

I tested this on a group that has 2 secor instances. I edited the conf & restarted one instance at a time. Is that wrong? Another approach would be to stop all instances first, make conf changes on all, and only then start all again.

----

After the change what we see is:

Offsets on zookeeper:

Topic | Partition | Offset | logSize | Lag | Created | Last Seen
-- | -- | -- | -- | -- |  -- | --
audit.ads |   | 10237750057 | 10237823723 | 73666 |   |   |  
  | 0 | 2587551247 | 2587569392 | 18145 |  6 months ago | 9 minutes ago
  | 1 | 2589058112 | 2589065679 | 7567 |  6 months ago | 4 minutes ago
  | 2 | 2529974789 | 2529999425 | 24636 |  6 months ago | 11 minutes ago
  | 3 | 2531165909 | 2531189227 | 23318 |  6 months ago | 10 minutes ago

Offsets on kafka:

Topic | Partition | Offset | logSize | Lag | Created | Last Seen
-- | -- | -- | -- | -- | --  | --
audit.ads |   | 10237750057 | 10237823799 | 73742 |  |  
  | 0 | 2587551247 | 2587569416 | 18169 | 9 minutes ago | 9 minutes ago
  | 1 | 2589058112 | 2589065680 | 7568 | 4 minutes ago | 4 minutes ago
  | 2 | 2529974789 | 2529999454 | 24665 |  11 minutes ago | 11 minutes ago
  | 3 | 2531165909 | 2531189249 | 23340 |  10 minutes ago | 10 minutes ago

They match, as expected.

But now output folder is missing the expected uploads:

```
s3smoke ls s3://ds-analytics-raw-smoke/secor/audit.ads/processdate=20181031/hour=14/
2018-10-31 16:40:38     685812 1_0_00000000002587527768.gz
2018-10-31 16:09:35     681064 1_1_00000000002588932703.gz
2018-10-31 16:09:23     675959 1_2_00000000002529868976.gz
2018-10-31 16:40:14     735433 1_3_00000000002531140914.gz
```

It's more than 30 minutes since anything was uploaded to partitions 1 & 2, yet their offsets have been committed recently! What's committing those offsets if nothing is uploaded?

I can see this effect even more clearly with the most recent folders, where the offsets have been recently committed for a partition, but there's no data file in the recent output folder for that partition yet. Possibly there is also data loss (offset gap) between different output folders.

**Comment by HenryCaiHaiying** [comment]
This PR:
https://github.com/pinterest/secor/pull/450#pullrequestreview-170348487 is
going to change the offset storage, you can try that PR to see if that can
fix your problem

**Comment by juhoautio** [comment]
@HenryCaiHaiying thanks, so does that PR support seamlessly migrating from zk to kafka as the offset storage without losing any data? Sorry, I read the comments in the PR but based on them I wouldn't be sure. Does it make secor initially read the offsets from zk but then switch to kafka? How secor know to do that? What is the migration process like – do I need to set some specific conf, should I first make sure that all instances are stopped, etc.? I'm willing to test it as long as I know how :)

**Comment by meticulo3366** [comment]
is this issue finally resolved?

**Comment by HenryCaiHaiying** [comment]
I haven't tracked this issue, there are several fixes on kafka 2.0.0 profile later.  I think you can give it a try.
